### ISSUE TITLE
GameGuru MAX - Storyboard Editor HUD Grid

### ISSUE THREAD

**Issue by MonkeyFrogStudio** [Dark-Basic-Software-Limited/GameGuruRepo#4331]
MAX is a great program, but it does some odd things. Here's an example of something I find odd and would like to see fixed:

![grid](https://user-images.githubusercontent.com/36735507/228013653-686496bd-1f8b-4328-891a-1ae06351b15d.png)

The grid for the Storyboard Editor's HUD is not based on a square, but is a rectangle. That's just ... odd. How can snapping to 1 in the vertical direction be a different length than snapping to 1 in the horizontal direction? Also, why is a grid setting of 1 so big? What is the grid snapping number based on? In other words, what does "1" mean? It's not 1 pixel. It's not 1mm. So, what does that number "1" represent?

To get nice HUDs with precise alignment, I think this grid needs to be improved. Since we're working with images, it might be nice to base grid snapping on the pixel, like most art programs do. And, please, the grids need to be square.

**Comment by davetgc** [comment]
Not a bug, so listed as an enhancement, but I agree, it's not what would be expected.

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC @plemsoft Can we revisit this 'enhancement' at some point? A rectangular grid is very strange and is unexpected by most users, I'd say. Thanks. :)

**Comment by plemsoft** [comment]
@MonkeyFrogStudio The HUD system is using percent system to place everything , that's why it looks rectangular. I do not think we can use a pixel system , as the placement should be the same on all resolution, well I could change it to fit the ratio but then vertical and horizontal would not be using the same GRID value , so that would also be strange ????? :)

**Comment by MonkeyFrogStudio** [comment]
I think the HUD screen is a representation of what it will look like in game. It's obviously not the size of my actual monitor's resolution since there are the MAX UI elements all around the HUD placement area. Perhaps what the HUD system should do is 'lock' elements to the top-right, top-left, bottom-right, bottom-left, center, center-bottom, center-top, etc. For really large displays, then perhaps there could be a setting (accessed by the player when playing the game) to increase the size of HUD elements. 

Despite all the screen resolutions out there, they primarily fall into a few categories - standard, wide, and ultrawide. It might be good to be able to design for all of these. For example, you could click "standard" and see a basic rectangle that represents all standard screen resolutions. You place your HUD elements accordingly. The same for 'wide' and another for 'ultrawide'. Then, when the game starts, if it detects the game is in one of the standard screen sizes, it defaults to that HUD design. Etc. 

As far as the grid ... all I can say is it's somehow off-putting to be rectangular and it's difficult to set up snapping to work properly as it's a bit meaningless if the grid in one direction is a different spacing then the grid in another.

**Comment by LeeBamberTGC** [comment]
@MonkeyFrogStudio @plemsoft As Preben says the grid respects the underlying percentage coordinate system so that it can work on any aspect ratio, the rectangle is just a visual abstraction of the fact most resolutions are wider than they are higher.  We can discuss it here before implementing any changes, so we do this once :)  As it stands, it's by design and correct, but given we have 2 puzzled expressions here, we can seek a solution that satisfies all.

I think we can certainly expand the tooltip from "Set Grid Size" to "Set the grid size as a percentage of the overall screen height", so the 10 becomes = each slot is 10% of the overall screen height.

The percentage system IS pretty capable, and we CAN use fractions of a percent if we choose.  How about a new tick-box "Use Square Grid", ticked by default that resolves the grid into squares by adding extra snap points along the axis that is wider, so that the actual grid size becomes a square.  A simple example would be a resolution of 2000x1000 and a grid size of 10% of the height would split the grid into:

Vertical Slots = 1000/10 = 100 pixels per grid square = 10 grid slots on the vertical
Horizontal Slots = 2000/(100 pixels from above calc) = 20 grid slots on the horizontal

This would produce a screen with a grid 20x10 and appears as square slots.

As the percentage system uses a float value, using a real-world aspect ratio, say 1920x1080, you would get:

Vertical Slots = 1080/10 = 108 pixels per grid square = 10 grid slots on the vertical
Horizontal Slots = 1920/(108 pixels from above calc) = 17.77 grid slots on the horizontal

Again, the grid would be around 17x10 and be seen as exact squares.

A further refinement might be that we adjust the percentage spacing so that our horizontal grid slots land on a full integer, so:

Adjusted Horizontal Slots = 17.77 truncated to 17.
Adjusted Percentage Spacing = 100 / 17 = 5.882% per grid width

This would produce a nearly perfect square grid, but one that exactly fits into the screen, and the 5.882 percent snapping happens in the background with floats accurate enough to maintain this grid.

**Comment by plemsoft** [comment]
@MonkeyFrogStudio @LeeBamberTGC "Use Square Grid" sound like a good way to do this , added it but not on by default, but its stored in prefs so it pretty easy to enable by default, if everything works as intended we can enable it by default :)

**Comment by MonkeyFrogStudio** [comment]
Here's the issue as I see it - a grid for design work is made to accurately position pieces relative to each other, not just the 'page' (in this case, the screen). Let's say I have several pieces to a HUD element that sits in the lower-right corner of the screen. I would need to position each of these elements so they look good. This could mean aligning to the left edge of another element, to the right edge, centering it relative to another element, etc. With the current grid system, this is very difficult to do in many cases as you cannot set up a pixel perfect grid to snap elements to like this. 

For example, let's say I have a HUD element that's a frame. Then I want to place a number in this frame that for ammo and I want that centered in that frame. Then I want a 'health' bar below this frame that represents the 'health' of the weapon (let's say the weapon can get damaged with use) and I want that centered, too. And I have another HUD element on the lower right that needs to be aligned to the left edge of the frame, etc. If I could have a standard, square grid that I can size as needed (like in any other graphic design package) I could easily accomplish this. In MAX, with the current grid system for HUD elements, I cannot.

I would recommend that the grid be divorced from the screen resolution. 

Instead, I think the HUD creator should allow for three basic designs - standard, wide, and ultrawide. 

The end-user can reposition their HUD elements based on each screen. For example, I might want wide and ultrawide to also use the standard HUD setup, allowing HUD elements on the edges of Standard to NOT relatively in the same place in both wide and ultrawide. Or I might want it different for each one.

What I think MAX should do, considering there are several screen resolutions in each (standard, wide, and ultrawide) is take what position we give each element and set them relatively via a percentage. 10% from edge or whatever. Center. Etc. 

Basically, I see the grid as a tool for accurately placing HUD elements relative to each other.

**Comment by LeeBamberTGC** [comment]
@MonkeyFrogStudio OK, sounds like this feature is expanding beyond a square grid request, so reverting back to me for further discussions.  Does your vision for this require that the grid be aligned to actual pixels, or can it be done with screen percentages? Leaving the left/right/top alignment and sub-HUD-element alignment alone for a second (as this overlaps another feature request related to HUD groups for better element placement), just focusing on the grid on/off and grid size, what is missing when it comes to GRID?

**Comment by MonkeyFrogStudio** [comment]
Hi @LeeBamberTGC ,

Please understand that any requests I'm making are only requests. They certainly don't have to be implemented. :)

So, yes, aligned to ACTUAL PIXELS seems best. Why? Because the people who will be creating the actual HUD graphics will do so in a graphics program that uses actual pixels. A HUD element may be 800px by 200px, for example. And if I have another that's 100px by 50px and want to center that over the top of the 800px by 200px HUD element, I need to be able to set up a grid so I can easily do that. If the grid in the HUD designer were to use actual pixels, then it would be a no-brainer for artists moving from Photoshop, Clip Studio Paint, Paint Tool SIA, Krita, GIMP to create the individual HUD elements and get them to look the same in MAX.

And, yes, HUD groups would be perfect! That way all the HUD elements for the top-left could be a GROUP. Just wanted to add my 2 pence in for that one. ;)

**Comment by LeeBamberTGC** [comment]
@MonkeyFrogStudio No worries, these are all great suggestions and provides for the kind of control you would want when designing your game screens.  Naturally, overall layouts (i.e., top left vs bottom right vs bottom center) would be ultimately informed by aspect ratio and screen resolution, but the 'grouping' of HUD elements would benefit from having some kind of pixel-level awareness to match the actual art being imported, To kick off some small experiements, do you have time to knock up a quick and dirty HUD panel that absolutely requires precise pixel-level locking of content, for example:

HUD PANEL BOX 200x100 = Contains a center slot inside it located at 90,40 -> 110,60
HUD ICON 20x20 = "NEEDS" to be placed at 90,40 or it will look wrong
HUD TITLE 200x10 = Must be located at 0,0 for it to look right

Something like the above would all be grouped together and then the whole composite can be dragged around and locked to different alignment zones, or can be locked to the current grid snapping alignments.

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC 

Sure, I can crank out something very, very basic for you. Please give me a day or two since I am doing some other work.

One thing to mention, especially since I've seen other games handle this differently:

On wide screens, especially ultrawide, it may be beneficial to allow different settings for wide and ultrawide than for standard screen resolutions. With ultrawide screens, it can be very difficult to see HUD elements if they are way off in the corners (top-right/left and bottom right/left). My screen resolution, for example, is 5120 x 1440. If I am playing a game where my character has health and that health bar is way off at the far top-left, I have to take my eyes off combat to see where my health is at. Because of this, some games place this HUD element at the top corner for standard display resolutions (and maybe even wide), but for ultrawide, the HUD element is at the top, but NOT at the far corner. It's more like they took a standard or wide screen resolution and told the HUD element to go no farther even if the screen is wider. 

Personally, I think it's important to let the developer of the game decide how they want their HUD elements to appear on each type of screen - standard, wide, and ultrawide.
